presto-spi/src/main/java/com/facebook/presto/spi/block/SliceArrayBlock.java

```diff
@@ -201,25 +208,31 @@ public int getRegionSizeInBytes(int positionOffset, int length)
             throw new IndexOutOfBoundsException("Invalid position " + positionOffset + " in block with " + positionCount + " positions");
         }
 
-        long sizeInBytes = length * (long) (ARRAY_OBJECT_INDEX_SCALE + SLICE_INSTANCE_SIZE);
-        for (int i = positionOffset; i < positionOffset + length; i++) {
-            Slice value = values[i];
-            if (value != null) {
-                sizeInBytes += value.length();
-            }
-        }
-        return intSaturatedCast(sizeInBytes);
+        return getSliceArraySizeInBytes(values, positionOffset, length);
     }
 
     static int getSliceArrayRetainedSizeInBytes(Slice[] values)
     {
-        long sizeInBytes = SizeOf.sizeOf(values);
+        long sizeInBytes = sizeOf(values);
         Map<Object, Boolean> uniqueRetained = new IdentityHashMap<>(values.length);
+        boolean emptySliceExists = false;
         for (Slice value : values) {
-            if (value != null && value.getBase() != null && uniqueRetained.put(value.getBase(), true) == null) {
-                sizeInBytes += value.getRetainedSize();
+            if (value != null) {
+                if (value.getBase() != null) {
+                    if (uniqueRetained.put(value.getBase(), true) == null) {
+                        sizeInBytes += value.getRetainedSize();
+                    }
+                }
+                else if (value.equals(EMPTY_SLICE)) {
+                    emptySliceExists = true;
+                }
             }
         }
-        return intSaturatedCast(sizeInBytes);
+        return intSaturatedCast(sizeInBytes) + (emptySliceExists ? EMPTY_SLICE.getRetainedSize() : 0);
+    }
+
+    private static int getDistinctSliceArrayRetainedSizeInBytes(int sliceArraySizeInBytes, int nSlices)
```

Comment:
We don't generally name variable this way: `nSlices`

Please inline this function since it's only used once and it's pretty hard to come up with a good name for this function here.

If line 59 becomes too long as a result, we can rewrite it with `if`.